Add Apprise notification integration

caronc opened the pull request:
## Description:
Added notification support for [Apprise](https://github.com/caronc/apprise) which drastically increases the number off notification services supported by Home Assistant.  It also, unfortunately, provides some over-lap to services already supported.

## Docs PR:
https://github.com/home-assistant/home-assistant.io/pull/10451

## Example entry for `configuration.yaml` (if applicable):

```yaml
# Example configuration.yaml entry using Apprise URLs
notify:
  - name: NOTIFIER_NAME
    platform: apprise
    url: Define one ore more Apprise URLs
```

Configuration files can also be sourced (via another location on the same server, or a remote server.  Apprise supports [TEXT](https://github.com/caronc/apprise/wiki/config_text) and [YAML](https://github.com/caronc/apprise/wiki/config_yaml) based configuration formats.
```yaml
# Example configuration.yaml entry using Apprise Configuration URLs
notify:
  - name: NOTIFIER_NAME
    platform: apprise
    config: Define one ore more Apprise Configuration URLs
```

Apprise URLs loaded directly (using the `url` directive) or through configuration files are all appended together in one list.  You can use both `config` and `url` in the same configuration if you wish.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

caronc committed:
Added apprise notification component

homeassistant commented:
Hi @caronc, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#26868).

Once you do that we will be able to review and accept this pull request.

Thanks!

marchingphoenix commented on homeassistant/components/apprise/notify.py:
```diff
@@ -0,0 +1,71 @@
+def get_service(hass, config, discovery_info=None):
+    """Get the Apprise notification service."""
+    from apprise import Apprise
+    from apprise import AppriseConfig
```
could be combined
`from apprise import Apprise, Appriseconfig`

caronc replied:
Thanks; just pushed an update to accommodate this now.

caronc committed:
flake-8 fixes; black formatting + import merged to 1 line

caronc committed:
pylint issues resolved

frenck commented on homeassistant/components/apprise/manifest.json:
```diff
@@ -0,0 +1,10 @@
+    "apprise==0.8.0"
+  ],
+  "dependencies": [],
+  "codeowners": []
```
Thank you for your contribution thus far! 🎖 Since this is a new contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

Please, add your GitHub username to the `manifest.json` of this integration.

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

caronc replied:
Thank you; I've added my name to the `manifest.json` as requested.

frenck commented on homeassistant/components/apprise/notify.py:
```diff
@@ -0,0 +1,69 @@
+
+def get_service(hass, config, discovery_info=None):
+    """Get the Apprise notification service."""
+    from apprise import Apprise, AppriseConfig
```
Please move this import to the top of the file

caronc replied:
Done.  It was originally placed in `get_service()` to follow the same structure as other notification services (such as pushbullet, discord, etc).  By keeping it on top of the file, don't you risk throwing an import error every time this file is referenced (even for those who don't plan on using it)?

Anyway; i still made the change :slightly_smiling_face:  as requested.

frenck replied:
It used to be the other way around, however, not all has been migrated yet.

caronc committed:
added github name to manifest.json

caronc committed:
import moved to top as per code review request

caronc committed:
manifest formatting to avoid failing ci

caronc committed:
.coveragerc updated to include apprise

frenck commented on .coveragerc:
```diff
@@ -36,6 +36,7 @@ omit =
     homeassistant/components/apache_kafka/*
     homeassistant/components/apcupsd/*
     homeassistant/components/apple_tv/*
+    homeassistant/components/apprise/*
```
You've provided tests, why exclude it from coverage?

caronc replied:
Doh! I have no excuse; I think i got caught up of searching for other components and just copy/pasting their examples. I'll fix this tonight; good catch.

caronc committed:
removed block for written tests

caronc committed:
more test coverage

MartinHjelmare commented on homeassistant/components/apprise/notify.py:
```diff
@@ -0,0 +1,72 @@
+
+    loaded = len(a_obj)
+    if not loaded:
+        _LOGGER.error("No Apprise services were loaded.")
```
We don't end logging messages with period.

caronc replied:
Fixed

MartinHjelmare commented on homeassistant/components/apprise/notify.py:
```diff
@@ -0,0 +1,72 @@
+        _LOGGER.error("No Apprise services were loaded.")
+        return None
+
+    _LOGGER.info("Loaded %d Apprise service(s).", loaded)
```
Debug.

caronc replied:
Fixed.

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,107 @@
+from tests.common import assert_setup_component, get_test_home_assistant
+
+
+class TestApprise(unittest.TestCase):
```
New tests must be standalone pytest test functions. See eg:

https://github.com/home-assistant/home-assistant/blob/1c72a246a048d0f2f3d3dca0b77be559dae01a78/tests/components/manual/test_alarm_control_panel.py#L83-L119

caronc replied:
@MartinHjelmare, first off: thank you for your help offline.  I was able to build onto the tests and get much further as a result.  I'm still hitting a wall, but I do see a light at the end of the tunnel at least.

The test that is failing is: [test_apprise_notification()](https://github.com/home-assistant/home-assistant/blob/cdf482eb88eb9e3d556e56b3893fcd58edb3985b/tests/components/apprise/test_notify.py#L52).  It appears that `hass.services.async_call()` is causing the tests to abort.

The error:
```
E    homeassistant.exceptions.ServiceNotFound: Unable to find service notify/test

../../homeassistant/core.py:1209: ServiceNotFound
```

Your thoughts/advice would be most appreciated (or anyone for that matter reviewing this). :slightly_smiling_face:

MartinHjelmare replied:
We need to set up the correct component. See my change suggestions below.

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,107 @@
+        self.hass = get_test_home_assistant()
+        self.tmp_dir = mkdtemp()
+
+    def tearDown(self):  # pylint: disable=invalid-name
```
Use pytest fixtures with `yield` for set up and teardown code.

caronc committed:
formatting as per code review

caronc committed:
tests converted to async style as per code review

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+
+    config = {"apprise": {"name": "test", "platform": "apprise", "config": "/path/"}}
+
+    with patch("apprise.AppriseConfig.add", return_value=False):
```
Be careful to patch in the correct place.
https://docs.python.org/3/library/unittest.mock.html#where-to-patch

I think we should change this to:
```suggestion
    with patch("homeassistant.components.apprise.notify.AppriseConfig.add", return_value=False):
```

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+        assert await async_setup_component(hass, BASE_COMPONENT, config)
+        await hass.async_block_till_done()
+
+    with patch("apprise.AppriseConfig.add", return_value=True):
```
```suggestion
    with patch("homeassistant.components.apprise.notify.AppriseConfig.add", return_value=True),
```

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+        await hass.async_block_till_done()
+
+    with patch("apprise.AppriseConfig.add", return_value=True):
+        with patch("apprise.Apprise.add", return_value=False):
```
```suggestion
        patch("homeassistant.components.apprise.notify.Apprise.add", return_value=False):
```

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+            "url": "mailto://user:pass@example.com",
+        }
+    }
+    with patch("apprise.Apprise.add", return_value=False):
```
```suggestion
    with patch("homeassistant.components.apprise.notify.Apprise.add", return_value=False):
```

caronc replied:
It's working as is, but would you still prefer i change this (see comment below)

MartinHjelmare replied:
Ok.

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+    data = {"title": "Test Title", "message": "Test Message"}
+    with patch("apprise.Apprise.notify", return_value=True):
```
```suggestion
    with patch("homeassistant.components.apprise.notify.Apprise.notify", return_value=True):
```

caronc replied:
The adaptation for Apprise into Home Assistant doesn't really require a lot of testing since 99.999% of the testing is already covered in Apprise (which has 100% testing coverage). My initial goal for test cases were to just pass through all of the sections of new code I added.  there is no magic taking place otherwise to test.

On a side note (for this specific test): Strangely the **patch()** works fantastic as is on everything but the **Apprise.notify()** function. It executes the command in both the old way and your new suggested way.  It seemed to work when i just mocked the entire **Apprise()** which is what the latest merge has. Let me know if you still want me to explicitly access the imported module instead of the general library (both achieve the same results).

MartinHjelmare commented:
Please don't squash commits after review has started to make it easier for readers to track changes.

caronc commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+            await hass.async_block_till_done()
+
+
+async def test_apprise_config_load_okay(hass, tmpdir):
```
Done :)

caronc commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+    with patch("apprise.AppriseConfig.add", return_value=False):
+        assert await async_setup_component(hass, BASE_COMPONENT, config)
+        await hass.async_block_till_done()
+
```
That's the grey area. Apprise does all of the grunt work that normally otherwise sits in your `component/*` directories.  There isn't much to test since it's [already handle](https://codecov.io/github/caronc/apprise).  The tests I wrote just test the responses and attempt to access all portions of the new code i added to Home Assistant.  I can see with the merge i did last night i missed one.  I'll revisit that soon and do another push.

MartinHjelmare replied:
I think we can check if the service for this platform has been registered.

caronc replied:
Done!

caronc commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,68 @@
+async def test_apprise_config_load_fail(hass):
+    """Test apprise configuration failures."""
+
+    config = {"apprise": {"name": "test", "platform": "apprise", "config": "/path/"}}
```
Thank you; it's working great now!

caronc committed:
increased coverage

caronc commented:
Just wanted to apologize, I attempted to rebase with upstream/dev, and push the changes, but instead it just made it so I changed 299+ files.  Changes reverted, but I don't know how to undo the review request from all 15 of you... :flushed:

caronc committed:
bumped version of apprise to 0.8.1

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,85 @@
+
+        # Test the call to our underlining notify() call
+        await hass.services.async_call(BASE_COMPONENT, "test", data)
+        await hass.async_block_till_done()
```
Here we should assert that the library notify method has been called.

caronc replied:
Done :slightly_smiling_face:

caronc committed:
test that mocked entries are called

caronc committed:
added tests for hass.service loading

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -76,10 +77,19 @@
 
     with patch("apprise.Apprise") as mock_apprise:
-        mock_apprise.notify.return_value = True
+        obj = MagicMock()
```
I suggest we rename the variable to something more specific, eg `apprise_obj`.

caronc replied:
Done!

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,110 @@
+        await hass.async_block_till_done()
+
+        # Test that our service failed to load
+        assert hass.services.has_service(BASE_COMPONENT, "test") is False
```
```suggestion
        assert not hass.services.has_service(BASE_COMPONENT, "test")
```

caronc replied:
Done

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,110 @@
+            await hass.async_block_till_done()
+
+            # Test that our service failed to load
+            assert hass.services.has_service(BASE_COMPONENT, "test") is False
```
```suggestion
            assert not hass.services.has_service(BASE_COMPONENT, "test")
```

caronc replied:
I applied your suggestion to all references :slightly_smiling_face: .  This one is also resolved as the last identified below.

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,110 @@
+    await hass.async_block_till_done()
+
+    # Valid configuration was loaded; our service is good
+    assert hass.services.has_service(BASE_COMPONENT, "test") is True
```
```suggestion
    assert hass.services.has_service(BASE_COMPONENT, "test")
```

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,110 @@
+        await hass.async_block_till_done()
+
+        # Test that our service failed to load
+        assert hass.services.has_service(BASE_COMPONENT, "test") is False
```
```suggestion
        assert not hass.services.has_service(BASE_COMPONENT, "test")
```

MartinHjelmare commented on tests/components/apprise/test_notify.py:
```diff
@@ -0,0 +1,110 @@
+        await hass.async_block_till_done()
+
+        # Test the existance of our service
+        assert hass.services.has_service(BASE_COMPONENT, "test") is True
```
```suggestion
        assert hass.services.has_service(BASE_COMPONENT, "test")
```

caronc committed:
support tags for those who identify the TARGET option

caronc committed:
renamed variable as per code review

caronc committed:
'assert not' used instead of 'is False'

MartinHjelmare commented on homeassistant/components/apprise/notify.py:
```diff
@@ -0,0 +1,73 @@
+
+        If no target/tags are specified, then services are notified as is
+        However, if any tags are specified, then they will be applied
+        to the notification causing filtering (if set up that way)
```
The linter may complain about missing period at the end.

caronc replied:
Added a period _just in case_ :slightly_smiling_face:

MartinHjelmare reviewed (approved):
Looks good!

Can be merged when build passes.

caronc committed:
added period (in case linter isn't happy)

caronc commented:
@MartinHjelmare Thanks for your help and patients in getting the testing working with me.  I've updated the docs here in this PR: https://github.com/home-assistant/home-assistant.io/pull/10451 to reference the `TARGET` reference i added.

Quick question though:  In Apprise it's referenced as *tagging*.  Do you prefer use of the `target` reference for consistency across all notification services? or is it okay that I use `tag` for consistency with Apprise?  This is in reference to the actual sending of a notification [here](https://github.com/home-assistant/home-assistant/blob/16b5ace39573177c62f1344796601a102a72d801/homeassistant/components/apprise/notify.py#L71).

Thoughts?

MartinHjelmare commented:
We have to use the target argument since that is what the service call may provide.

MartinHjelmare commented:
Great!
